power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -400,8 +438,22 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
     // get sequence idx map of an entire batch for fast caching of component sequences
     // (only applicable for independent update dataset)
     SequenceIdx get_sequence_idx_map(ConstDataset const& update_data) const {
-        assert(is_update_independent(update_data));
-        return get_sequence_idx_map(update_data, 0);
+        auto update_components_independence = check_components_independence(update_data);
+        assert(std::ranges::all_of(update_components_independence,
+                                   [](auto const& comp) { return !comp.has_id || comp.ids_match; }));
+
+        // TODO: (jguo) this function could be encapsulated in UpdateCompIndependence in update.hpp
```

Comment:
👍

Reply:
Refactor scoped out. Noted in https://github.com/PowerGridModel/power-grid-model/issues/806